File: Phase1_Modulations/channel/add_awgn_noise.m
function [rxSig, noiseVar, snrDb] = add_awgn_noise(txSig, EbNoDb, bitsPerSymbol)
%ADD_AWGN_NOISE  Add complex AWGN from Eb/N0.

txSig = txSig(:);

% Calculates the average power of the transmitted complex signal
txPower = mean(abs(txSig).^2);

% Converts Energy per Bit (Eb/N0) to Signal-to-Noise Ratio (SNR) by 
% factoring in the spectral efficiency (bits per symbol)
snrDb = EbNoDb + 10*log10(bitsPerSymbol);

% Converts logarithmic dB into a linear scale
snrLin = 10^(snrDb/10);

% Determines the required noise variance based on the desired SNR
noiseVar = txPower / snrLin;

% Generates complex Gaussian noise. The variance is divided by 2 because 
% the noise power must be split equally between the real (In-Phase) and 
% imaginary (Quadrature) components
noise = sqrt(noiseVar/2) * (randn(size(txSig)) + 1i*randn(size(txSig)));

% Add the noise vector to the transmitted signal vector
rxSig = txSig + noise;
end

File: Phase1_Modulations/channel/add_awgn_noise.py
import numpy as np


def add_awgn_noise(tx_sig, EbNo_db, bits_per_symbol):
    tx_sig = np.asarray(tx_sig, dtype=np.complex128).flatten()
    tx_power = np.mean(np.abs(tx_sig) ** 2)
    snr_db = EbNo_db + 10 * np.log10(bits_per_symbol)
    snr_lin = 10 ** (snr_db / 10)
    noise_var = tx_power / snr_lin
    noise = np.sqrt(noise_var / 2) * (
        np.random.randn(tx_sig.size) + 1j * np.random.randn(tx_sig.size)
    )
    rx_sig = tx_sig + noise
    return rx_sig, noise_var, snr_db
